src/wallet/asyncrpcoperation_mergetoaddress.h

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2017 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef ASYNCRPCOPERATION_MERGETOADDRESS_H
+#define ASYNCRPCOPERATION_MERGETOADDRESS_H
+
+#include "amount.h"
```

Comment:
Attaching here as comment:

Funds were merged into a different zaddr (in the wallet), and not the zaddr (in the wallet) specified.  This was run on mainnet.

```
./zcash-cli z_mergetoaddress ANY_ZADDR zcdYqZzp5ABTtsKi8PdEZefwjWfbeuBMNRn91pWcLsbZAztDcUBXwx25aaGk1NoSCUP66AA5VMBhv9kCQE3V1zTMYUFRERy 0.00001
{
  "remainingUTXOs": 0,
  "remainingTransparentValue": 0.00000000,
  "remainingNotes": 0,
  "remainingShieldedValue": 0.00000000,
  "mergingUTXOs": 0,
  "mergingTransparentValue": 0.00000000,
  "mergingNotes": 1,
  "mergingShieldedValue": 0.03872239,
  "opid": "opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b"
}
```

debug.log:

```
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: z_mergetoaddress initialized (params={"fromaddresses":"ANY_ZADDR","toaddress":"zcdYqZzp5ABTtsKi8PdEZefwjWfbeuBMNRn91pWcLsbZAztDcUBXwx25aaGk1NoSCUP66AA5VMBhv9kCQE3V1zTMYUFRERy","fee":0.00001000})
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: spending 0.03872239 to send 0.03871239 with fee 0.00001
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: transparent input: 0.00
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: private input: 0.03872239
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: private output: 0.03871239
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: fee: 0.00001
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: spending note (txid=1404b71a86, vjoinsplit=0, ciphertext=0, amount=0.03872239, height=260958, confirmations=10)
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: generating note for change (amount=0.03871239)
2018-01-26 00:54:58 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: creating joinsplit at index 0 (vpub_old=0.00, vpub_new=0.00001, in[0]=0.03872239, in[1]=0.00, out[0]=0.00, out[1]=0.03871239)
...
2018-01-26 00:55:35 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: Payment Disclosure: js=0, n=1, zaddr=zcDEb4i1nYPHgAABg86TuhoBLJ1aSTpA8kK4bE32eXcjaJdABKymaG5YAWfgiyGCk9mZocXQFHjBeuGkBFpAGqAv5ERDtYx
2018-01-26 00:55:35 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: Payment Disclosure: js=0, n=0, zaddr=zcUxZKkUd6TwWtxYxJta2A4kYkz4qHoyEdJBiB4FvyKTheYdjHNNJUR73qMmFZw4rqwbTjwLxjhYMBkgwQxVWKPqawmULqp
...
2018-01-26 00:55:35 opid-77f1dc74-3513-41a3-b4f2-88625b4cff4b: z_mergetoaddress finished (status=success, txid=0123d0c0a003cabd3eb7460a73e78d258f0d665c6904d224100d25249cd1f12e)
```

Payment disclosure shows index 0 using an address in the wallet, whereas index 1 is a dummy address.

```
Result is that funds found their way to a different zaddr in wallet, as shown by index 0 of js.  

./zcash-cli z_getbalance zcUxZKkUd6TwWtxYxJta2A4kYkz4qHoyEdJBiB4FvyKTheYdjHNNJUR73qMmFZw4rqwbTjwLxjhYMBkgwQxVWKPqawmULqp
0.03871239
```

The test above was to an address in the same wallet.  Have not tried with an address in a different wallet.

Reply:
Added a case to the RPC test and replicated the bug. Seems like the to address is being ignored in this case, and all funds are being returned to the change address (see the `generating note for change` line in the trace).

Reply:
Simple bug - I had forgotten to alter the target address of the final change output, so it was going to the z-address for the first note (which gets used as the change address for intermediate change).